etna/lib/etna/route.rb

```diff
@@ -124,7 +124,13 @@ def process_call(app, request)
       update_params(request)
 
       unless authorized?(request)
-        return cc_redirect(request) if cc_available?(request)
+        if cc_available?(request)
+          if request.content_type == 'application/json'
+            return [ 403, { 'Content-Type' => 'application/json' }, [ { error: 'You are forbidden from performing this action, but you can visit the project home page and request access.' }.to_json ] ]
```

Comment:
api calls should still just 403 here, since there is not a browser context to request guest access with.

Reply:
Ah, that makes sense, since it'll be from an XHR request.

Does this mean that the client libraries in `etna-js` will have to do the right messaging or redirects to the code of conduct page? For example, if someone opens Timur and from the root view clicks into IPI, I think the first time they'll hit an endpoint that triggers this code is when the client app does a fetch from Magma `/retrieve` ... ? So would that have to catch the 403 and then redirect the app? (same problem whether it's the 302 or the 403)

Reply:
Not quite -- check timur/server.rb:

```
    with auth: { user: { can_view?: :project_name } } do

...
      # remaining view routes are parsed by the client and must also be set there
      get ':project_name', as: :project do
        erb_view(:client)
      end

      get ':project_name/*view_path', as: :project_view do
        erb_view(:client)
      en
```

So in other words, even the document views will trigger this from the browser.

Reply:
Ah, nice, I missed the top of the block when I was scanning :facepalm: 

:+1:

Reply:
Hmm, Janus server.rb currently has: `get '/*views' do erb_view(:client) end`; the project-specific page is unguarded and wouldn't be caught this way, they would just show up and get a broken/empty page.

Reply:
Right, so I need to swap out the 'get project' endpoint for 'get projects' on the cc view, that way they can still pull open projects without permission and see the page correctly, making the change.